lib/grppcaut.gi

```diff
@@ -869,12 +907,319 @@ end;
+    act:=BaseSteinitzVectors(IdentityMat(n,f),
+	  bas);
+    act:=Concatenation(bas,act.factorspace);
+    stb:=[];
+    # subspace
+    for i in GeneratorsOfGroup(GL(l,f)) do
+      m:=IdentityMat(n,f);
+      m{[1..l]}{[1..l]}:=i;
+      Add(stb,m);
+    od;
+    # factor
+    for i in GeneratorsOfGroup(GL(n-l,f)) do
+      m:=IdentityMat(n,f);
+      m{[l+1..n]}{[l+1..n]}:=i;
+      Add(stb,m);
+    od;
+    # mixer
+    m:=IdentityMat(n,f);
+    m[l+1][1]:=One(f);
+    Add(stb,m);
+    stb:=List(stb,x->x^act); # base change
+    stb:=Group(stb);
+    SetSize(stb,Size(GL(l,f))*Size(GL(n-l,f))*Size(f)^(l*(n-l)));
+    Info(InfoMatOrb,2,"reduced full GL by ",Size(G)/Size(stb));
+    return stb;
+  fi;
+  if not OrbitIsLonger(GeneratorsOfGroup(G),bas,OnSubspacesByCanonicalBasis,
+	   Size(f)^QuoInt(Length(bas[1]),2)) then
+    #just plain
+    act:=OnSubspacesByCanonicalBasis;
+    xset:=Orbit(G,bas,act);
+    stb:=[Position(xset,bas)];
+  else
+    act:=OnLines;
+    stb:=NormedVectors(VectorSpace(f,bas));
+    xset:=Union(Orbits(G,stb,act));
+    stb:=Set(List(stb,x->Position(xset,x)));
+  fi;
+  hom:=ActionHomomorphism(G,xset,act,"surjective");
+  act:=PreImage(hom,Stabilizer(Image(hom),stb,OnSets));
+  Info(InfoMatOrb,2,"reduce ",Size(G), " to ",Size(act));
+  return act;
+end);
+
+# construct subgroup of GL that stabilizes the spaces given and fixes teh
+# listed spaceorbits.
+
+# auxillary
+RedmatSpanningIndices:=function(gens)
+local bas,n,one,new,a,b,g;
+  n:=Length(gens[1]);
+  one:=One(gens[1]);
+  new:=[];
+  bas:=[];
+  while Length(bas)<n do
+    a:=First(one,x->Length(bas)=0 or SolutionMat(bas,x)=fail);
+    Add(new,Position(one,a));
+    Add(bas,a);
+    # spin
+    for b in bas do
+      for g in gens do
+        a:=b*g;
+	if SolutionMat(bas,a)=fail then
+	  Add(bas,a);
+	fi;
+      od;
+    od;
+  od;
+  return new;
+end;
+
+SpaceAndOrbitStabilizer:=function(n,field,spaces,osporb)
+local spaceincl,outvecs,l,sub,yet,i,j,k,s,t,new,incl,min,rans,sofar,done,
+      gens,one,spl,ngens,m,sz,a,sporb,notyet;
+
+  spaceincl:=function(big,small)
+    return RankMat(Concatenation(big,small))=Length(big);
+  end;
+
+  outvecs:=function(space,new)
+    return BaseSteinitzVectors(new,space).factorspace;
+  end;
+
+  one:=IdentityMat(n,field);
+  sub:=[]; # space so far
+  spaces:=Unique(List(spaces,x->OnSubspacesByCanonicalBasis(x,one)));
+
+  SortBy(spaces,Length);
+  if not ForAny(spaces,x->Length(x)=n) then
+    Add(spaces,List(one,ShallowCopy));
+  fi;
+
+  osporb:=List(osporb,x->List(x,x->OnSubspacesByCanonicalBasis(x,one)));
+  sporb:=List(osporb,ShallowCopy);
+  l:=Length(spaces);
+
+  # inclusion relation, tests
+  repeat
+    yet:=0;
+    done:=[0];
+    incl:=List([1..l],x->[0,x]);
+    new:=[];
+    for i in [1..l] do
+      for j in [i+1..l] do
+	if spaceincl(spaces[j],spaces[i]) then
+	  AddSet(incl,[i,j]);
+	else
+	  # check for meet/join closed
+	  s:=SumIntersectionMat(spaces[i],spaces[j]);
+	  for k in s do
```

Comment:
Both `SumIntersectionMat` and `RankMat` (whcih is what `spaceincl` calls) end up using `SemiEchelonMatDestructive`. So we compute it twice in the worst case! Let's not do that :-). The easiest (albeit not most efficient, though hopefully it won't mattter) way woud be to always compute `SumIntersectionMat`, and then check if `s[2] = spaces[i]`.